### PR TITLE
Fancy new rust trait for a Martian stage.

### PR DESCRIPTION
**Pull request by pmarks**
Strongly typed! Less boilerplate! Tests!

### CODE REVIEW

File: src/stage.rs [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+use serde::Serialize;
+use Metadata;
+use utils::{obj_decode, obj_encode};
+use types::{MartianMakePath, MartianVoid};
+
+/// Memory/ thread request can be negative in matrian 
+/// http://martian-lang.org/advanced-features/#resource-consumption
+#[derive(Debug, Serialize, Deserialize, Copy, Clone, Default)]
+pub struct Resource {
+    #[serde(rename = "__mem_gb")]
```
[the remark this answers is not in the document]
Comment (sreenathkrishnan) [reply src/stage.rs]: It is possible to write a custom serializer to achieve that, but I don't understand why that is needed here. Could you elaborate?

[the remark this answers is not in the document]
Comment (adam-azarchs) [reply src/stage.rs]: Just to reduce noise.  Actually serializing a request for 0 threads or memory would be confusing.

[the remark this answers is not in the document]
Comment (sreenathkrishnan) [reply src/stage.rs]: The default value for the resource would be `None`. The expectation is that the user would set it to a non-zero value if needed. Does 0 threads or memory mean anything special for martian?

[the remark this answers is not in the document]
Comment (adam-azarchs) [reply src/stage.rs]: martian is written in go, which just assigns zero values when json keys are missing.

File: src/stage.rs [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+    mem_gb: Option<isize>,
+    #[serde(rename = "__threads")]
+    threads: Option<isize>,
```
[the remark this answers is not in the document]
Comment (sreenathkrishnan) [reply src/stage.rs]: Added `__vmem_gb`

File: src/stage.rs [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+
+    pub fn set_join_resource(&mut self, join_resource: Resource) {
+        self.join_resource = join_resource;
+    }
+}
+
+pub struct MartianRover {
+    files_path: PathBuf,
+    mem_gb: usize,
+    threads: usize,
```
[the remark this answers is not in the document]
Comment (sreenathkrishnan) [reply src/stage.rs]: Done!

**Pull request merged**